### PR TITLE
Update Memo structure to latest specification

### LINKED ISSUE #199: Refactor Memo to follow the specification
`Memo` should be

```
memo = [ + (tstr / (bstr .cbor any)) ]
```

See specification in [1]

[1]: https://github.com/many-protocol/specification/blob/f1a2bad7ee9413d4ec6c068660e2366e05c342e6/spec/cddl/types.cddl#L41-L47

### PR DESCRIPTION
**Pull request by hansl**
Fixes #199

### CODE REVIEW

**Comment by hansl** [comment]
Explanation of the latest 2 commits; for some reason `cargo install` started failing and I had to upgrade the rust toolchain to 1.63 (I think `cargo install` will use latest stable?). Then new clippy errors started popping so I had to fix those.

File: src/many-modules/src/_9_account/features/multisig.rs
Comment (fmorency) [inline src/many-modules/src/_9_account/features/multisig.rs:153]: Isn't that already checked by the `MemoInner`, i.e., can't you simply do

```
Ok(Self::from(MemoInner::try_from(b)?))
```

File: src/many-modules/src/_9_account/features/multisig.rs
Comment (fmorency) [inline src/many-modules/src/_9_account/features/multisig.rs:102]: Would this allow the creation of a `Memo` with an empty `inner`?

We need to enforce that `The memo itself CANNOT be empty and MUST contain at least one entry.`

**Review by fmorency — COMMENTED** [review]
Thanks!

Some questions before approval.

File: src/many-modules/src/_9_account/features/multisig.rs
Comment (hansl) [reply src/many-modules/src/_9_account/features/multisig.rs:153]: You're right. I did the `String` version but not this one. My bad.

File: src/many-modules/src/_9_account/features/multisig.rs
Comment (hansl) [reply src/many-modules/src/_9_account/features/multisig.rs:102]: Removed `Default` and `new()`.

**Review by fmorency — APPROVED** [review]

**Review by Denommus — APPROVED** [review]

**Comment by hansl** [comment]
@fmorency if you have time, I added a backward compatible decoding so we can make a migration in framework.

**Comment by hansl** [comment]
I'll block merging this PR for now (moving to Draft) so that we can merge it when the migration in framework is ready.

**Comment by fmorency** [comment]
> @fmorency if you have time, I added a backward compatible decoding so we can make a migration in framework.

I will look at it after the migration in framework is ready.

File: src/many-modules/src/_9_account/features/multisig.rs
Comment (Denommus) [inline src/many-modules/src/_9_account/features/multisig.rs:131]: Why is this a `TryFrom` and not a `From` if there's no failure path?

File: src/many-modules/src/_9_account/features/multisig.rs
Comment (Denommus) [inline src/many-modules/src/_9_account/features/multisig.rs:96]: This is wrong. Memo is a `[ + (tstr / (bstr .cbor any)) ]`, which means there must be at least one element. You need another field `first: MemoInner<MAX_LENGTH>` and adapt the code accordingly. This way you're preventing an empty `Memo` in the type level.

File: src/many-modules/src/_9_account/features/multisig.rs
Comment (Denommus) [inline src/many-modules/src/_9_account/features/multisig.rs:176]: It's possible to decode an empty `Memo` from here, which, as previously stated, is a wrong behavior according to the specification.

**Comment by hansl** [comment]
Blocked on https://github.com/liftedinit/many-framework/pull/243 (because we need migrations).

File: src/many-modules/src/_9_account/features/multisig.rs
Comment (hansl) [reply src/many-modules/src/_9_account/features/multisig.rs]: Done and added a test.

File: src/many-modules/src/_9_account/features/multisig.rs
Comment (hansl) [reply src/many-modules/src/_9_account/features/multisig.rs]: I'm not going to add a new type for `NeverEmptyVec` (or add a crate). It's a simple invariant to test for.

**Comment by Denommus** [comment]
CI failing

File: src/many-modules/Cargo.toml
Comment (fmorency) [inline src/many-modules/Cargo.toml:24]:
```suggestion
cbor-diag = "0.1.11"
```

File: src/many-modules/src/_4_events.rs
Comment (fmorency) [inline src/many-modules/src/_4_events.rs:964]: nit: Impl sane `defaults()`

File: src/many-types/Cargo.toml
Comment (fmorency) [inline src/many-types/Cargo.toml:22]:
```suggestion
cbor-diag = "0.1.11"
```

File: src/many-types/src/memo/legacy.rs
Comment (fmorency) [inline src/many-types/src/memo/legacy.rs:66]: I don't think this code should not be located in this file

File: src/many-types/src/memo.rs
Comment (fmorency) [inline src/many-types/src/memo.rs:209]: Check the integrity of the data

File: src/many-types/src/memo.rs
Comment (fmorency) [inline src/many-types/src/memo.rs:223]: Check the integrity of the data

File: src/many-types/src/memo.rs
Comment (fmorency) [inline src/many-types/src/memo.rs:236]: Check the integrity of the data

File: src/many-types/src/memo.rs
Comment (fmorency) [inline src/many-types/src/memo.rs:245]: Check the error type

File: src/many-types/src/memo.rs
Comment (fmorency) [inline src/many-types/src/memo.rs:253]: Check the error type

File: src/many-types/src/memo.rs
Comment (fmorency) [inline src/many-types/src/memo.rs:262]: Check the integrity of the data

File: src/many-types/src/memo.rs
Comment (fmorency) [inline src/many-types/src/memo.rs:257]:
```suggestion
    fn bytes_decode_ok() {
```
nit

File: src/many-types/src/memo.rs
Comment (fmorency) [inline src/many-types/src/memo.rs:266]:
```suggestion
    fn bytes_decode_large() {
```
nit

File: src/many-types/src/memo.rs
Comment (fmorency) [inline src/many-types/src/memo.rs:271]: Check the error type

File: src/many-types/src/memo.rs
Comment (fmorency) [inline src/many-types/src/memo.rs:280]: Check the integrity of the data

File: src/many-types/src/memo.rs
Comment (fmorency) [inline src/many-types/src/memo.rs:289]: Check the error type

File: src/many-types/src/memo.rs
Comment (fmorency) [inline src/many-types/src/memo.rs:284]:
```suggestion
    fn mixed_decode_bytes_too_lare() {
```

File: src/many-types/src/memo.rs
Comment (fmorency) [inline src/many-types/src/memo.rs:293]:
```suggestion
    fn mixed_decode_memo_type_mismatch() {
```
nit

File: src/many-types/src/memo.rs
Comment (fmorency) [inline src/many-types/src/memo.rs:297]: Check the error type

File: src/many-types/src/memo.rs
Comment (fmorency) [inline src/many-types/src/memo.rs:301]:
```suggestion
    fn mixed_decode_memo_type_mismatch_array() {
```
nit

File: src/many-types/src/memo.rs
Comment (fmorency) [inline src/many-types/src/memo.rs:305]: Check the error type

**Review by fmorency — CHANGES_REQUESTED** [review]
Minor stuff. LGTM otherwise! Thanks!

File: src/many-modules/src/_4_events.rs
Comment (hansl) [reply src/many-modules/src/_4_events.rs:964]: I don't want to be gloomy, but `default()` is how block 9400 came to be :/ I'm tempted to remove it from everywhere until we have a solid solution.

File: src/many-types/src/memo/legacy.rs
Comment (hansl) [reply src/many-types/src/memo/legacy.rs]: The `Data` type was mixed with the old `Memo` type, so they're tied together. The new `Memo` type is essentially the merge of both of those two fields.

**Review by hansl — COMMENTED** [review]
PTAL.

**Review by fmorency — APPROVED** [review]
Lgtm

File: src/many-types/src/memo.rs
Comment (fmorency) [inline src/many-types/src/memo.rs:145]: Do we really need this check?

**Review by fmorency — APPROVED** [review]
LGTM assuming my question is addressed.

My macro-fu is unable to review this piece, but it looks like the others.

File: src/many-types/src/memo.rs
Comment (hansl) [reply src/many-types/src/memo.rs:145]: Yes. If there's more than 1 string, `== "some string"` should always be false.

File: src/many-migration/src/lib.rs
Comment (fmorency) [inline src/many-migration/src/lib.rs:215]: nit: rename this method, it's confusing

File: src/many-migration/src/lib.rs
Comment (fmorency) [inline src/many-migration/src/lib.rs:217]: `==` instead of `>`?

File: src/many-migration/src/lib.rs
Comment (fmorency) [inline src/many-migration/src/lib.rs:153]: `>`

File: src/many-migration/src/lib.rs
Comment (fmorency) [inline src/many-migration/src/lib.rs:411]: As discussed on Slack

File: src/many-migration/src/lib.rs
Comment (fmorency) [inline src/many-migration/src/lib.rs:438]: As discussed on Slack

**Review by fmorency — CHANGES_REQUESTED** [review]
Minor stuff. Else LGTM

File: src/many-migration/src/lib.rs
Comment (hansl) [reply src/many-migration/src/lib.rs:242]: How did that past the tests :/ Need to test more.

File: src/many-migration/src/lib.rs
Comment (hansl) [reply src/many-migration/src/lib.rs:166]: Nope, when it's ==.

File: src/many-migration/src/lib.rs
Comment (hansl) [reply src/many-migration/src/lib.rs:446]: That's the current height. This function is deprecated anyway.

File: src/many-migration/src/lib.rs
Comment (fmorency) [reply src/many-migration/src/lib.rs:195]: This comment is still wrong

File: src/many-migration/src/lib.rs
Comment (fmorency) [inline src/many-migration/src/lib.rs:405]: Why are you keeping height here?

You don't use it anywhere but in `self.height = block_height;`

File: src/many-migration/src/lib.rs
Comment (fmorency) [inline src/many-migration/src/lib.rs:375]: Please test serialization/deserialization of `MigrationConfig`

File: src/many-migration/src/lib.rs
Comment (fmorency) [inline src/many-migration/src/lib.rs:189]: Modify this message to be consistent with the others.

File: src/many-migration/src/lib.rs
Comment (fmorency) [inline src/many-migration/src/lib.rs:380]: Test those

File: src/many-migration/src/lib.rs
Comment (fmorency) [inline src/many-migration/src/lib.rs:483]: Test those

File: src/many-migration/src/lib.rs
Comment (fmorency) [inline src/many-migration/src/lib.rs:453]: This line doesn't seem tested

**Review by fmorency — CHANGES_REQUESTED** [review]
Some questions/changes.

Do you have time to add more tests to memo, e.g., try_from(Either), Encode(bstr), try_from_iter(), push_str, push_bytevec, etc

File: src/many-migration/src/lib.rs
Comment (hansl) [reply src/many-migration/src/lib.rs:405]: Removed.

File: src/many-migration/src/lib.rs
Comment (hansl) [reply src/many-migration/src/lib.rs:189]: Actually the others are wrong, fixed.

File: src/many-migration/src/lib.rs
Comment (hansl) [reply src/many-migration/src/lib.rs:375]: really?

Comment (fmorency) [reply src/many-migration/src/lib.rs:375]: Perhaps not. You seem to test it in fw anyway

Comment (fmorency) [reply src/many-migration/src/lib.rs:375]: Yes, I agree

**Comment by hansl** [comment]
I hope this is enough ;)

image

**Review by fmorency — APPROVED** [review]
Assuming CI is fixed

**Comment by fmorency** [comment]
Please log an issue with the remaining stuff to do/fix

**Pull request merged**
